client/src/components/exercise/lab10/components/ShapeSpawner.js

```diff
@@ -0,0 +1,356 @@
+  return (
+    (objectPosition + IMG_SIZE >= x && objectPosition <= x) ||
+    (objectPosition >= x && objectPosition + IMG_SIZE <= x + size) ||
+    (objectPosition <= x + size && objectPosition + IMG_SIZE >= x + size)
+  );
+};
+
+/**
+ * True if the Moving Object is overlapping with a Falling Shape on the Y Plane.
+ */
+const isTouchingY = (height, y, size) => {
+  return y + size >= height - IMG_SIZE;
+};
+
+/**
+ * ShapeSpawner component responsible for the spawning of shapes
+ * Be cautious when updating logic here
+ */
+const ShapeSpawner = (props) => {
+  const [shapes, setShapes] = useState([]);
+  const fpsRef = useRef(60);
+  const intervalRef = useRef(null);
+  const requestRef = useRef(null);
+
+  /**
+   * Update reference value when prop value is updated
+   */
+  useEffect(() => {
+    if (props.fps !== fpsRef.current) {
+      fpsRef.current = props.fps;
+    }
+  }, [props.fps]);
+
+  /**
+   * Dependency on Shapes for whenever a Y value is updated.
+   * Dependency on Object Position for whenever its X value is updated.
+   */
+  useEffect(() => {
+    const newShapes = shapes.filter(({ x, y, size, color, empty }) => {
+      if (empty) {
+        return true;
+      }
+
+      const touchingX = isTouchingX(props.objectPosition, x, size);
+      const touchingY = isTouchingY(
+        props?.parentRef?.current?.offsetHeight,
+        y,
+        size
+      );
+      const shapesCollided = touchingX && touchingY;
+
+      if (shapesCollided && props.collectWeights) {
+        props.actions.updateColorWeight(color);
+      }
+
+      return !shapesCollided;
+    });
+
+    !_.isEqual(newShapes, shapes) && setShapes(newShapes);
+  }, [shapes, props.objectPosition]);
+
+  /**
+   * AI Logic.
+   * Moves towards an empty slot if one exist.
+   * Otherwise, do not touch the color with the heaviest weight.
+   */
+  useEffect(() => {
+    const withinBounds = (leftShape, rightShape) => {
+      if (leftShape === null) {
+        return props.objectPosition + IMG_SIZE < rightShape.x;
+      } else if (rightShape === null) {
+        return props.objectPosition > leftShape.x + leftShape.size;
+      } else {
+        return (
+          props.objectPosition + IMG_SIZE < rightShape.x &&
+          props.objectPosition > leftShape.x + leftShape.size
+        );
+      }
+    };
+
+    if (props.simulationStatus === SIMULATION_STARTED && props.ai) {
+      let minDifference = Infinity;
+      let closestGapIndex = null;
+      for (let i = 0; i < shapes.length; i++) {
+        const shape = shapes[i];
+        if (shape.empty) {
+          const absoluteDifference = Math.abs(props.objectPosition - shape.x);
+          if (absoluteDifference < minDifference) {
+            minDifference = absoluteDifference;
+            closestGapIndex = i;
+          }
+        }
+      }
+
+      if (!_.isEmpty(shapes)) {
+        /* If there is a Shape at every column */
+        if (closestGapIndex === null) {
```

Comment:
is this for the bias part you are planning to implement later?

Reply:
yes